### ISSUE TITLE
Hyperlane DefiLlama Adapter

### ISSUE THREAD

**Issue by nambrot** [hyperlane-xyz/hyperlane-monorepo#4396]
Hyperlane is currently not listed on popular sites like defillama (https://defillama.com/)

This bounty is to add an integration that allows defillama to properly show Hyperlane data.

Relevant links: 
- https://github.com/DefiLlama/bridges-server

This integration should probably read the relevant artifacts for warp routes and mailboxes from https://github.com/hyperlane-xyz/hyperlane-registry

**Comment by raizo07** [comment]
@nambrot I want to work on this. Kindly assign me

**Comment by nambrot** [comment]
Go for it!

**Comment by jiuxme** [comment]
This isn't a bad idea

**Comment by tmsrjs** [comment]
Hey @nambrot, how's it going?
It took longer than expected but I managed to get this working using the registry, although I did need to make some minor changes to their helpers to allow for dynamically built adapters. Let's hope they can incorporate them, otherwise the adapter would be a lot more verbose.
I went ahead and made it fully generic, meaning that it will track all transactions where there's either a `SentTransferRemote` or `ReceivedTransferRemote` event and also include a `Dispatch` and `Process` event from the corresponding mailbox.
Let me know if this works for you or you'd rather track only the warp routes in the registry.
Here's the PR for you to check https://github.com/DefiLlama/bridges-server/pull/287
Thanks!

**Comment by nambrot** [comment]
I think generic makes sense to me, how does defillama know not double count events on both origin and destination? @yorhodes calls out that a `SentTransferRemote` is not necessarily a deposit, and in general the isDeposit: bool might not be expressive enough to account for warp route configurations. (but not too familiar with the defillama abstraction)

**Comment by tmsrjs** [comment]
From my understanding by looking at other adapters, they consider a deposit an outgoing cross-chain transfer and a withdrawal (`isDeposit === false`) an incoming one. Assuming that's correct I see a direct correlation from `SentTransferRemote` to `isDeposit === true` and `ReceivedTransferRemote` to `isDeposit === false`, regardless of the kind of Hyp token since they all emit the same events, but I could very well be mistaken.  @yorhodes does this sound correct to you?
Regarding double count of events: I had the exact same doubt and by the looks of it they don't keep track of that since they do not expect the adapter to generate any kind of id to match the events on both chains.

**Comment by nambrot** [comment]
I was under the impression that only a deposit of collateral (i.e. hyperc20 collateral or hypnative counts), but tbh we could just submit it this way and ask the defillama team? It does look like CCTP does it that way since there is no collateral ever being deposited

**Comment by tmsrjs** [comment]
yep, that's one of the adapters I looked at to arrive at that conclusion.
The PR is already submitted, guess we'll just have to wait until someone from the DefiLlama team takes a look at it.
